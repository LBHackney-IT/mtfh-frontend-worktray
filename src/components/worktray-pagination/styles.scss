.lbh-pagination {
  &__link {
    &__no-item {
      color: black !important;
      font-weight: unset !important;
    }
  }
}
