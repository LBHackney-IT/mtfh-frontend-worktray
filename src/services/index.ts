export { default as locale } from "./locale";
export { config } from "./config";
