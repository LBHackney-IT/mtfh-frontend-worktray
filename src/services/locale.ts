const locale = {
  title: "Your tasks",
  views: {
    worktray: {
      noWorktrayResults: "No tasks found",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  components: {
    filters: {
      applyFilters: "Apply filters",
      clearFilters: "Clear filters",
      selectAll: "Select All",
      removeAll: "Remove All",
    },
    pagination: {
      previous: "Previous",
      next: "Next",
      assistiveNavigation: (page: number): string => `Page ${page}`,
      searchResultsCounter: (page: number, pageSize: number, total: number): string =>
        total === 0
          ? "Showing 0 of 0 results"
          : `Showing ${page * pageSize - pageSize + 1}â€”${Math.min(
              page * pageSize,
              total,
            )} of ${total} results`,
    },
    controls: {
      timePeriodOptions: {
        "30": "Last 30 days",
        "60": "Last 60 days",
        "3": "Last 3 months",
        "6": "Last 6 months",
        "12": "Last 12 months",
      },
      limitOption: (value: number): string => `${value} items`,
    },
  },
};

export default locale;
