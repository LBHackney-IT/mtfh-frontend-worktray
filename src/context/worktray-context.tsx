import React, {
  Dispatch,
  FC,
  Reducer,
  createContext,
  useContext,
  useEffect,
  useMemo,
  useReducer,
} from "react";
import { useHistory, useLocation } from "react-router-dom";

import { stringify } from "query-string";

import { AxiosSWRError, useAxiosSWR } from "@mtfh/common/lib/http";

import { config } from "../services";
import {
  LimitOptions,
  OrderByOptions,
  Process,
  ProcessSortOptions,
  TimePeriodOptions,
  WorktrayFilterOptions,
} from "../types";

interface WorktrayResults {
  results: {
    processes: Process[];
  };
  total: number;
}

export type WorktrayState = {
  page: number;
  pageSize: LimitOptions;
  timePeriod: TimePeriodOptions;
  sort: ProcessSortOptions;
  order: OrderByOptions;
  patch?: string;
  areaId?: string | null;
  processNames?: string;
  status?: string;
  results?: Process[];
  total?: number;
  error?: AxiosSWRError;
};

export type WorktrayActions =
  | {
      type: "PAGE";
      payload: number;
    }
  | {
      type: "LIMIT";
      payload: LimitOptions;
    }
  | {
      type: "TIME_PERIOD";
      payload: TimePeriodOptions;
    }
  | {
      type: "SORT";
      payload: ProcessSortOptions;
    }
  | {
      type: "FILTER";
      payload: {
        type: WorktrayFilterOptions;
        payload: string;
      };
    }
  | {
      type: "ORDER";
      payload: OrderByOptions;
    };

export interface WorktrayProviderProps {
  initial: Partial<WorktrayState>;
}

const getInitialState = (
  options: WorktrayProviderProps["initial"] = {},
): WorktrayState => {
  return {
    page: options.page || 1,
    pageSize: options.pageSize || LimitOptions.SMALL,
    timePeriod: options.timePeriod || TimePeriodOptions.DAYS_30,
    sort: options.sort || ProcessSortOptions.STATUS,
    order: options.order || OrderByOptions.ASC,
    processNames: options.processNames || undefined,
    patch: options.patch || undefined,
    areaId: options.areaId || undefined,
    status: options.status || undefined,
  };
};

export const WorktrayContext = createContext<{
  state: WorktrayState;
  dispatch: Dispatch<WorktrayActions>;
}>({
  state: getInitialState(),
  dispatch: () => {},
});

export const WorktrayProvider: FC<WorktrayProviderProps> = ({ children, initial }) => {
  const reducer: Reducer<WorktrayState, WorktrayActions> = (state, action) => {
    switch (action.type) {
      case "PAGE": {
        return {
          ...state,
          page: action.payload,
        };
      }
      case "LIMIT": {
        if (Object.values(LimitOptions).includes(action.payload)) {
          return {
            ...state,
            pageSize: action.payload,
          };
        }
        return state;
      }
      case "TIME_PERIOD": {
        if (Object.values(TimePeriodOptions).includes(Number(action.payload))) {
          return {
            ...state,
            timePeriod: action.payload,
          };
        }
        return state;
      }
      case "SORT": {
        if (Object.values(ProcessSortOptions).includes(action.payload)) {
          return {
            ...state,
            sort: action.payload,
          };
        }
        return state;
      }
      case "ORDER": {
        if (Object.values(OrderByOptions).includes(action.payload)) {
          return {
            ...state,
            order: action.payload,
          };
        }
        return state;
      }
      case "FILTER": {
        return {
          ...state,
          [action.payload.type]: action.payload.payload,
        };
      }
      default: {
        return state;
      }
    }
  };

  const initialState = getInitialState(initial);
  const [state, dispatch] = useReducer(reducer, initialState);
  const { ...query } = state;

  const { data, error } = useAxiosSWR<WorktrayResults>(
    `${config.searchApiUrl}/search/processes?${stringify(
      {
        searchText: query.patch,
        page: query.page,
        pageSize: query.pageSize,
        timePeriod: query.timePeriod,
        processNames: query.processNames,
        status: query.status,
        sortBy: query.sort,
        isDesc: query.order === "asc",
      },
      {
        skipEmptyString: true,
      },
    )}`,
    {
      refreshWhenHidden: false,
      revalidateOnFocus: false,
    },
  );

  useEffect(() => {
    if (data?.total && data.total > 0) {
      const totalPages = Math.ceil(data.total / state.pageSize);
      if (state.page > totalPages) {
        dispatch({ type: "PAGE", payload: totalPages });
      }
    }
  }, [state.page, data?.total, state.pageSize]);

  const value = useMemo(
    () => ({
      state: {
        ...state,
        results: data?.results.processes,
        total: data?.total,
        error,
      },
      dispatch,
    }),
    [state, data, error, dispatch],
  );

  return <WorktrayContext.Provider value={value}>{children}</WorktrayContext.Provider>;
};

export const PersistWorktrayContextURL = ({
  sessionKey,
}: {
  sessionKey?: string;
}): null => {
  const {
    state: { page, pageSize, order, sort, timePeriod, patch, processNames, status },
  } = useContext(WorktrayContext);
  const { push } = useHistory();

  useEffect(() => {
    const query = new URLSearchParams();
    if (page > 1) {
      query.set("p", String(page));
    }

    if (pageSize !== LimitOptions.SMALL) {
      query.set("l", String(pageSize));
    }

    query.set("t", String(timePeriod));

    if (order && order !== OrderByOptions.ASC) {
      query.set("o", order);
    }

    query.set("sort", String(sort));

    if (patch) {
      query.set("patch", patch);
    }

    if (processNames) {
      query.set("processNames", processNames);
    }

    if (status) {
      query.set("status", status);
    }

    const path = `?${query.toString()}`;
    push(path);

    if (sessionKey) {
      window.sessionStorage.setItem(sessionKey, path);
    }
  }, [
    page,
    pageSize,
    order,
    sort,
    sessionKey,
    push,
    timePeriod,
    patch,
    processNames,
    status,
  ]);

  return null;
};

export const WorktrayURLProvider: FC<{
  sessionKey?: string;
  patchId?: string | null;
  areaId?: string | null;
}> = ({ children, sessionKey, patchId, areaId }): JSX.Element => {
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  const page = Number(query.get("p") || 1);
  const pageSize = Number(query.get("l") || 10);
  const timePeriod = (query.get("t") || "") as TimePeriodOptions;
  const patch = query.get("patch") || patchId || "";
  const processNames = query.get("processNames") || "";
  const status = query.get("status") || "";
  const sortBy = query.get("sort");
  const orderBy = query.get("o");

  const sort = useMemo(() => {
    if (sortBy === "name") {
      return ProcessSortOptions.NAME;
    }
    return ProcessSortOptions.STATUS;
  }, [sortBy]);

  const order = useMemo(() => {
    if (orderBy === "desc") {
      return OrderByOptions.DESC;
    }
    return OrderByOptions.ASC;
  }, [orderBy]);

  return (
    <WorktrayProvider
      initial={{
        page,
        sort,
        order,
        pageSize,
        timePeriod,
        patch,
        areaId,
        processNames,
        status,
      }}
    >
      {children}
      <PersistWorktrayContextURL sessionKey={sessionKey} />
    </WorktrayProvider>
  );
};
