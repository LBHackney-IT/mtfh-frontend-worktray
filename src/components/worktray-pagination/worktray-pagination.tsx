import React, { useContext, useMemo } from "react";

import {
  Pagination,
  PaginationButton,
  PaginationControls,
  PaginationSummary,
} from "@mtfh/common/lib/components";

import { WorktrayContext } from "../../context/worktray-context";
import { locale } from "../../services";
import "./styles.scss";

const { previous, next, assistiveNavigation, searchResultsCounter } =
  locale.components.pagination;

interface WorktrayPaginationProps {
  pageRange?: number;
}

export const hasNoResult = (total: number | undefined): boolean => {
  return !total || total <= 0;
};

export const WorktrayPagination = ({
  pageRange = 2,
}: WorktrayPaginationProps): JSX.Element | null => {
  const {
    state: { total, page, pageSize },
    dispatch,
  } = useContext(WorktrayContext);

  const [totalPages, activePage] = useMemo(() => {
    const tp = total ? Math.ceil(total / pageSize) : 0;
    return [tp, page > tp ? tp : page];
  }, [total, page, pageSize]);

  const range = useMemo(() => {
    if (hasNoResult(total)) {
      return [1];
    }
    let rangeStart = activePage - pageRange;
    let rangeEnd = activePage + pageRange;
    if (rangeEnd > totalPages) {
      rangeEnd = totalPages;
      rangeStart = totalPages - pageRange * 2;
    }

    if (rangeStart <= 1) {
      rangeStart = 1;
      rangeEnd = Math.min(pageRange * 2 + 1, totalPages);
    }

    const length = rangeEnd - rangeStart + 1;
    const visiblePages = new Array<number>(length);

    for (let i = 0; i < length; i += 1) {
      visiblePages[i] = rangeStart + i;
    }

    return visiblePages;
  }, [total, activePage, pageRange, totalPages]);

  if (total === undefined || total < 0 || totalPages === Infinity || range.length === 0) {
    return null;
  }

  return (
    <Pagination role="navigation">
      <PaginationSummary>{searchResultsCounter(page, pageSize, total)}</PaginationSummary>
      <PaginationControls>
        {activePage > 1 && (
          <PaginationButton
            as="button"
            onClick={() => dispatch({ type: "PAGE", payload: activePage - 1 })}
            variant="previous"
          >
            {previous}
          </PaginationButton>
        )}
        {range.map((index) => (
          <PaginationButton
            as="button"
            key={index}
            onClick={() => dispatch({ type: "PAGE", payload: index })}
            active={hasNoResult(total)}
            aria-label={assistiveNavigation(index)}
            aria-current={index === activePage ? "page" : undefined}
            className={hasNoResult(total) ? "lbh-pagination__link__no-item" : undefined}
          >
            {index}
          </PaginationButton>
        ))}
        {activePage < totalPages && (
          <PaginationButton
            as="button"
            onClick={() => dispatch({ type: "PAGE", payload: activePage + 1 })}
            variant="next"
          >
            {next}
          </PaginationButton>
        )}
      </PaginationControls>
    </Pagination>
  );
};
