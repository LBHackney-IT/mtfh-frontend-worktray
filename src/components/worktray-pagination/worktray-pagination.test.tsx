import React from "react";

import { render, server } from "@hackney/mtfh-test-utils";
import { fireEvent, screen } from "@testing-library/react";
import { rest } from "msw";

import { WorktrayProvider } from "../../context/worktray-context";
import { locale } from "../../services";
import { WorktrayPagination } from "./worktray-pagination";

const getMockWorktrayResults = (data = {}) =>
  rest.get("/api/worktray", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        results: [],
        total: 40,
        ...data,
      }),
    );
  });

const pageNumbersCheck = () => {
  expect(screen.queryByText("2")).not.toBeInTheDocument();
  expect(screen.getByText("3")).toBeInTheDocument();
  expect(screen.getByText("4")).toBeInTheDocument();
  expect(screen.getByText("5")).toBeInTheDocument();
  expect(screen.getByText("6")).toBeInTheDocument();
  expect(screen.getByText("7")).toBeInTheDocument();
  expect(screen.queryByText("8")).not.toBeInTheDocument();
};

describe("worktray-pagination", () => {
  test("Pagination displays range from start if range exceeds start", async () => {
    server.use(getMockWorktrayResults());
    render(
      <WorktrayProvider initial={{}}>
        <WorktrayPagination />
      </WorktrayProvider>,
    );

    expect(
      await screen.findByText(locale.components.pagination.next),
    ).toBeInTheDocument();
    expect(
      screen.queryByText(locale.components.pagination.previous),
    ).not.toBeInTheDocument();
    expect(screen.getByText("1")).toBeInTheDocument();
    expect(screen.queryByText("5")).not.toBeInTheDocument();
  });

  test("Pagination displays range", async () => {
    server.use(getMockWorktrayResults({ total: 100 }));
    render(
      <WorktrayProvider initial={{ page: 5 }}>
        <WorktrayPagination />
      </WorktrayProvider>,
    );

    await expect(
      screen.findByText(locale.components.pagination.next),
    ).resolves.toBeInTheDocument();
    expect(screen.getByText(locale.components.pagination.previous)).toBeInTheDocument();
    pageNumbersCheck();
  });

  test("Pagination displays range from end if range exceeds end", async () => {
    server.use(getMockWorktrayResults({ total: 70 }));
    render(
      <WorktrayProvider initial={{ page: 7 }}>
        <WorktrayPagination />
      </WorktrayProvider>,
    );

    await expect(
      screen.findByText(locale.components.pagination.previous),
    ).resolves.toBeInTheDocument();
    expect(screen.queryByText(locale.components.pagination.next)).not.toBeInTheDocument();
    pageNumbersCheck();
  });

  test("Pagination wont display if the range is empty", async () => {
    server.use(getMockWorktrayResults({ total: 12 }));
    render(
      <WorktrayProvider initial={{ page: 1 }}>
        <WorktrayPagination />
      </WorktrayProvider>,
    );

    expect(screen.queryByText("1")).not.toBeInTheDocument();
  });

  test("Pagination navigation triggers state changes", async () => {
    server.use(getMockWorktrayResults());
    render(
      <WorktrayProvider initial={{ page: 1 }}>
        <WorktrayPagination />
      </WorktrayProvider>,
    );

    expect(await screen.findByText("1")).toHaveAttribute("aria-current", "page");

    fireEvent.click(screen.getByText(locale.components.pagination.next));
    expect(await screen.findByText("2")).toHaveAttribute("aria-current", "page");

    fireEvent.click(screen.getByText(locale.components.pagination.previous));
    expect(await screen.findByText("1")).toHaveAttribute("aria-current", "page");

    fireEvent.click(screen.getByText("3"));
    expect(await screen.findByText("3")).toHaveAttribute("aria-current", "page");
  });
});
