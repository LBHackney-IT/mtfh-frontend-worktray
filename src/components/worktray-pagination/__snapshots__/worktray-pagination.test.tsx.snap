// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worktray-pagination Pagination displays correctly for no results 1`] = `
<div>
  <nav
    class="lbh-pagination"
    role="navigation"
  >
    <div
      class="lbh-pagination__summary"
    >
      Showing 0 of 0 results
    </div>
    <ul
      class="lbh-pagination__list"
    >
      <li
        class="lbh-pagination__item"
      >
        <button
          aria-label="Page 1"
          class="lbh-pagination__link lbh-pagination__link--current lbh-pagination__link__no-item"
        >
          1
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
